File: Domain/Model/CodingSession.swift
import Foundation

/// Aggregate root representing a coding session with screenshots and analysis
class CodingSession {
    let sessionId: SessionId
    private(set) var screenshots: [Screenshot] = []
    private(set) var analysis: Analysis?

    init(sessionId: SessionId = SessionId()) {
        self.sessionId = sessionId
    }

    // MARK: - Business Logic

    /// Add a screenshot to the session
    /// - Parameter screenshot: The screenshot to add
    func addScreenshot(_ screenshot: Screenshot) {
        screenshots.append(screenshot)
    }

    /// Remove a screenshot by ID
    /// - Parameter screenshotId: The ID of the screenshot to remove
    /// - Returns: true if removed, false if not found
    @discardableResult
    func removeScreenshot(withId screenshotId: ScreenshotId) -> Bool {
        if let index = screenshots.firstIndex(where: { $0.id == screenshotId }) {
            screenshots.remove(at: index)
            return true
        }
        return false
    }

    /// Set the analysis result
    /// - Parameter analysis: The analysis to set
    func setAnalysis(_ analysis: Analysis) {
        self.analysis = analysis
    }

    /// Clear the entire session
    func clear() {
        screenshots.removeAll()
        analysis = nil
    }

    /// Clear only the analysis, keeping screenshots
    func clearAnalysis() {
        analysis = nil
    }

    // MARK: - Query Methods

    /// Check if session has screenshots
    var hasScreenshots: Bool {
        return !screenshots.isEmpty
    }

    /// Check if session has analysis
    var hasAnalysis: Bool {
        return analysis != nil
    }

    /// Get total number of screenshots
    var screenshotCount: Int {
        return screenshots.count
    }

    /// Get total size of all screenshots in KB
    var totalSizeInKB: Double {
        return screenshots.reduce(0) { $0 + $1.sizeInKB }
    }

    /// Check if session is ready for analysis (has at least one screenshot)
    var isReadyForAnalysis: Bool {
        return hasScreenshots
    }

    /// Get all screenshots as base64 strings for API
    var screenshotsAsBase64: [String] {
        return screenshots.map { $0.base64String }
    }
}

File: Domain/Model/CodingSession.cs
using System;
using System.Collections.Generic;
using System.Linq;
using InterviewMasterApp.Domain.Entities;
using InterviewMasterApp.Domain.ValueObjects;

namespace InterviewMasterApp.Domain.Model
{
    /// <summary>
    /// Aggregate root representing a coding session with screenshots and analysis.
    /// Migrated from Swift: CodingSession.swift
    /// </summary>
    public class CodingSession
    {
        public SessionId SessionId { get; }
        private List<Screenshot> _screenshots = new();
        private Analysis? _analysis;

        public CodingSession(SessionId? sessionId = null)
        {
            SessionId = sessionId ?? new SessionId();
        }

        // MARK: - Public Properties

        public IReadOnlyList<Screenshot> Screenshots => _screenshots.AsReadOnly();
        public Analysis? Analysis => _analysis;

        // MARK: - Business Logic

        /// <summary>
        /// Add a screenshot to the session.
        /// </summary>
        public void AddScreenshot(Screenshot screenshot)
        {
            if (screenshot == null)
                throw new ArgumentNullException(nameof(screenshot));

            _screenshots.Add(screenshot);
        }

        /// <summary>
        /// Remove a screenshot by ID.
        /// </summary>
        /// <returns>true if removed, false if not found</returns>
        public bool RemoveScreenshot(Guid screenshotId)
        {
            var screenshot = _screenshots.FirstOrDefault(s => s.Id == screenshotId);
            if (screenshot != null)
            {
                _screenshots.Remove(screenshot);
                return true;
            }
            return false;
        }

        /// <summary>
        /// Set the analysis result.
        /// </summary>
        public void SetAnalysis(Analysis analysis)
        {
            if (analysis == null)
                throw new ArgumentNullException(nameof(analysis));

            _analysis = analysis;
        }

        /// <summary>
        /// Clear the entire session.
        /// </summary>
        public void Clear()
        {
            _screenshots.Clear();
            _analysis = null;
        }

        /// <summary>
        /// Clear only the analysis, keeping screenshots.
        /// </summary>
        public void ClearAnalysis()
        {
            _analysis = null;
        }

        // MARK: - Query Methods

        /// <summary>
        /// Check if session has screenshots.
        /// </summary>
        public bool HasScreenshots => _screenshots.Count > 0;

        /// <summary>
        /// Check if session has analysis.
        /// </summary>
        public bool HasAnalysis => _analysis != null;

        /// <summary>
        /// Get total number of screenshots.
        /// </summary>
        public int ScreenshotCount => _screenshots.Count;

        /// <summary>
        /// Get total size of all screenshots in KB.
        /// </summary>
        public double TotalSizeInKb
        {
            get
            {
                return _screenshots.Sum(s => s.SizeInKb);
            }
        }

        /// <summary>
        /// Check if session is ready for analysis (has at least one screenshot).
        /// </summary>
        public bool IsReadyForAnalysis => HasScreenshots;

        /// <summary>
        /// Get all screenshots as base64 strings for API.
        /// </summary>
        public List<string> ScreenshotsAsBase64
        {
            get
            {
                return _screenshots
                    .Select(s => s.ToBase64())
                    .Where(b => b != null)
                    .ToList();
            }
        }

        public override string ToString()
        {
            return $"CodingSession(ID={SessionId}, Screenshots={ScreenshotCount}, " +
                   $"HasAnalysis={HasAnalysis})";
        }
    }
}
